File: 28.implement-strstr.go
/*
 * @lc app=leetcode id=28 lang=golang
 *
 * [28] Implement strStr()
 *
 * https://leetcode.com/problems/implement-strstr/description/
 *
 * algorithms
 * Easy (32.22%)
 * Total Accepted:    442.7K
 * Total Submissions: 1.4M
 * Testcase Example:  '"hello"\n"ll"'
 *
 * Implement strStr().
 * 
 * Return the index of the first occurrence of needle in haystack, or -1 if
 * needle is not part of haystack.
 * 
 * Example 1:
 * 
 * 
 * Input: haystack = "hello", needle = "ll"
 * Output: 2
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: haystack = "aaaaa", needle = "bba"
 * Output: -1
 * 
 * 
 * Clarification:
 * 
 * What should we return when needle is an empty string? This is a great
 * question to ask during an interview.
 * 
 * For the purpose of this problem, we will return 0 when needle is an empty
 * string. This is consistent to C's strstr() and Java's indexOf().
 * 
 */
func strStr(haystack string, needle string) int {
	if len(needle) == 0 {
		return 0
	}

	for i := 0; i <= len(haystack) - len(needle); i++ {
		found := true
		for j := 0; j<len(needle); j++ {
			if needle[j] != haystack[i+j] {
				found = false
				break
			}
		}
		if found {
			return i
		}
	}

	return -1
}

File: 2018/28.implement-strstr.cpp
/*
 * [28] Implement strStr()
 *
 * https://leetcode.com/problems/implement-strstr
 *
 * algorithms
 * Easy (28.05%)
 * Total Accepted:    198.5K
 * Total Submissions: 706.1K
 * Testcase Example:  '""\n""'
 *
 * 
 * Implement strStr().
 * 
 * 
 * Returns the index of the first occurrence of needle in haystack, or -1 if
 * needle is not part of haystack.
 * 
 */
class Solution {
public:
    int strStr(string haystack, string needle) {
		if(needle.empty()) {
			return 0;
		}

		for(int i=0; i<=(int)(haystack.size() - needle.size()); i++){
			if(haystack.substr(i, needle.size()) == needle){
				return i;
			}
		}

		return -1;
    }
};
